### PR TITLE
Add credential support to PCS API client

### PR DESCRIPTION
**Pull request by premun**
https://dev.azure.com/dnceng/internal/_workitems/edit/6423

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description
PCS client can now use Azure credentials for auth

### CODE REVIEW

File: src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        /// <summary>
+        /// Use this for darc invocations from services using an MI
+        /// </summary>
+        internal static PcsApiCredential CreateManagedIdentityCredential(string barApiBaseUri, string managedIdentityId)
+        {
+            string appId = EntraAppIds[barApiBaseUri.TrimEnd('/')];
+
+            var miCredential = new ManagedIdentityCredential(managedIdentityId);
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs:52]:
```suggestion
            ManagedIdentityCredential miCredential = new(managedIdentityId);
```

File: src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+            var appCredential = new ClientAssertionCredential(
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs:54]:
```suggestion
            ClientAssertionCredential appCredential = new(
```

File: src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            var requestContext = new TokenRequestContext(new string[] { $"api://{appId}/.default" });
+            return new PcsApiCredential(appCredential, requestContext);
+        }
+
+        /// <summary>
+        /// Use this for darc invocations from pipelines without a token.
+        /// </summary>
+        internal static PcsApiCredential CreateNonUserCredential(string barApiBaseUri)
+        {
+            var requestContext = new TokenRequestContext(new string[] { $"{EntraAppIds[barApiBaseUri.TrimEnd('/')]}/.default" });
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs:68]:
```suggestion
            TokenRequestContext requestContext = new(new string[] { $"{EntraAppIds[barApiBaseUri.TrimEnd('/')]}/.default" });
```

File: src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            var credential = new AzureCliCredential();
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Client/PcsApiCredential.cs:69]:
```suggestion
            AzureCliCredential credential = new();
```

**Review by dkurepa — DISMISSED** [review]

**Comment by premun** [comment]
@dkurepa the suggested changes won't work in this project because it uses C# 8.0

**Review by dkurepa — APPROVED** [review]

**Pull request merged**